### ISSUE TITLE
[CLOSED] Files not copying for independent template

### ISSUE THREAD

**Issue by schuhschuh** [cmake-basis/BASIS#238]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [ahundt](https://github.com/ahundt)**
_Tuesday Nov 12, 2013 at 20:41 GMT_
_Originally opened as https://github.com/schuhschuh/cmake-basis-legacy/issues/238_

---

I'm having trouble updating my template to use the feature/custom-doc-style branch. Aside from the problem with issue #231, I created a config.py for my template and it seems some of the files aren't copying. For example, I have doc/nrec_logo.png and doc/doxygen_extra.css, and neither of those files seem to copy, even when I specify --doc and --full as parameters. 

I use the following response file createProject.rsp:

```
   --name "nrecPractice"
   --author "Author Name <author@cmu.edu>" 
   --full 
   --description "This project is just to practice and test basis." 
   --template nrecBASISTemplate/nrecBASISTemplate-1.0
   --useopt OpenCV 
   --include 
   --src 
   --test 
   --doc
   --example

```

I included the config.py for my separate template here:

``` python
# project template configuration script for basisproject tool

# ------------------------------------------------------------------------------
version = 1.0

# ------------------------------------------------------------------------------
required = [
  # root documentation files
  'AUTHORS.txt',
  'README.txt',
  'INSTALL.txt',
  'COPYING.txt',
  'ChangeLog.txt',
  # root CMake configuration
  'CMakeLists.txt',
  'BasisProject.cmake'
]

# ------------------------------------------------------------------------------
options = {
  # additional configuration files
  'config-settings' : {
    'desc' : 'Include/exclude custom Settings.cmake file.',
    'path' : [ 'config/Settings.cmake' ]
  },
  'config-components' : {
    'desc' : 'Include/exclude custom Components.cmake file.',
    'path' : [ 'config/Components.cmake' ]
  },
  'config-package' : {
    'desc' : 'Include/exclude custom Package.cmake file.',
    'path' : [ 'config/Package.cmake' ]
  },
  'config-find' : {
    'desc' : 'Include/exclude custom Config.cmake.in file.',
    'path' : [
               'config/Config.cmake.in',
               'config/ConfigSettings.cmake'
             ]
  },
  'config-find-version' : {
    'desc' : 'Include/exclude custom ConfigVersion.cmake.in file.',
    'path' : 'config/ConfigVersion.cmake.in'
  },
  'config-script' : {
    'desc' : 'Include/exclude custom ScriptConfig.cmake.in file.',
    'path' : [ 'config/ScriptConfig.cmake.in' ]
  },
  'config-test' : {
    'desc' : 'Include/exclude custom CTestCustom.cmake.in file.',
    'path' : [ 'config/CTestCustom.cmake.in' ]
  },
  'config-use' : {
    'desc' : 'Include/exclude custom ConfigUse.cmake.in file.',
    'path' : [ 'config/ConfigUse.cmake.in' ]
  },
  'config' : {
    'desc' : 'Include/exclude all custom configuration files.',
    'deps' : [
               'config-settings',
               'config-components',
               'config-package',
               'config-find',
               'config-find-version',
               'config-script',
               'config-test',
               'config-use'
             ]
  },
  # software data
  'data' : {
    'desc' : 'Add/remove directory for auxiliary data files.',
    'path' : [ 'data/CMakeLists.txt' ]
  },
  # documentation
  'doc' : {
    'desc' : 'Add/remove directory for documentation files.',
    'path' : [ 
               'doc/CMakeLists.txt',
               'doc/doxygen_extra.css',
               'doc/doxygen_header.html.in',
               'doc/doxygen_footer.html.in',
               'doc/nrec_logo.png'
             ]
  },
  'doc-rst' : {
    'desc' : 'Add/remove reStructuredText (.rst) files for software manual/web site.',
    'path' : [
               'doc/CMakeLists.txt',
               'doc/index.rst',
               'doc/changelog.rst',
               'doc/download.rst',
               'doc/installation.rst',
               'doc/manual.rst',
               'doc/publications.rst',
               'doc/people.rst'
             ]
  },
  # usage example
  'example' : {
    'desc' : 'Add/remove directory for example files.',
    'path' : [ 'example/CMakeLists.txt' ]
  },
  # project modules
  'modules' : {
    'desc' : 'Add/remove support for modularization.',
    'path' : [ 'modules/' ]
  },
  # source files
  'include' : {
    'desc' : 'Add/remove directory for public header files.',
    'path' :[ 'include/' ]
  },
  'src' : {
    'desc' : 'Add/remove directory for project source files.',
    'path' : [ 'src/CMakeLists.txt' ]
  },
  # testing tree
  'test' : {
    'desc' : 'Add/remove support for testing.',
    'path' : [
               'CTestConfig.cmake',
               'test/CMakeLists.txt'
             ]
  },
  'test-internal' : {
    'desc' : 'Add/remove support for internal testing.',
    'path' : [ 'test/internal/CMakeLists.txt' ],
    'deps' : 'test'
  }
}

# ------------------------------------------------------------------------------
# presets
presets = {
  'minimal' : {
    'desc' : 'Choose minimal project template.',
    'args' : [ 'noconfig', 'nodata', 'nodoc', 'nodoc-rst', 'noexample', 'nomodules', 'noinclude', 'src' ]
  },
  'default' : {
    'desc' : 'Choose default project template.',
    'args' : [ 'noconfig', 'nodata', 'doc', 'doc-rst', 'noexample', 'nomodules', 'include', 'src', 'test' ]
  },
  'toplevel' : {
    'desc' : 'Create toplevel project.',
    'args' : [ 'noconfig', 'nodata', 'doc', 'doc-rst', 'noexample', 'modules', 'noinclude', 'nosrc', 'notest' ]
  },
  'module' : {
    'desc' : 'Create module of toplevel project.',
    'args' : [ 'noconfig', 'nodata', 'nodoc', 'nodoc-rst', 'noexample', 'nomodules', 'include', 'src', 'test' ]
  },
  'full' : {
    'desc' : 'Choose project template with all optional files.',
    'args' : [ 'config', 'doc', 'doc-rst', 'example', 'data', 'nomodules', 'include', 'src', 'test', 'test-internal' ]
  }
}

# ------------------------------------------------------------------------------
# additional substitutions besides <project>, <template-version>,...
from datetime import datetime as date

# ------------------------------------------------------------------------------
substitutions = {
  'date'  : date.today().strftime('%x'),
  'day'   : date.today().day,
  'month' : date.today().month,
  'year'  : date.today().year
}

del date
```

Also, would it be possible to put step by step instructions somewhere for configuring the default template to be different?

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Tuesday Nov 12, 2013 at 20:52 GMT_

---

Perhaps the problem is that specifying a template on the command line doesn't work correctly, because it loads the possible parameters from the default?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Tuesday Nov 12, 2013 at 23:37 GMT_

---

Sorry, my bad, it is indeed not considering the `--template` option.

Edit: Actually, it does. Just if no arguments are specified at all it used the default template for the help output. I will remove this, however, as displaying help for options which a particular template then does not actually have is not quite intuitive.

You mention `config.py`. Have you noticed that I renamed it to `_config.py` in between ?

Further, I should add checks to not read binary files such as images and try to alter them... not sure if that would be related, but at the moment it probably works only for text files properly.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:10 GMT_

---

I couldn't get markdown to do that first underscore :-)

I also realized I had a problem, and my test script was actually cloning an old version. However I'm still running into an issue with the command as follows, perhaps it is using the current working directory for some reason?

in ~/basis/test
○ ~/basis/test/basis-install/bin/basisproject --responsefile createProject.rsp
Could not find a BasisProject.cmake file in ~/basis/test! Are you sure this is a BASIS project?

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:13 GMT_

---

perhaps there should be a way to print and update the current default template as well?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:14 GMT_

---

Have you checked if the response file is actually being read ? I commented your code for the moment and wanted to enable later when I had time to actually take a closer look myself.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:16 GMT_

---

What do you mean by "print and update the current default template" ?

When you run `basisproject` with the `--verbose` option, it will tell you which directories it uses.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:26 GMT_

---

the way my response file code works at the moment is pretty simple, it just recursively runs the command line parser, finds any response files to open, then runs the command line parser on those until there aren't any left. Then you've basically constructed a list of all remaining parameters, and it runs on those. 

Inefficient but that doesn't matter since it should never have to go over enough files that it matters. There is the potential for an infinite loop though if two files reference each other.

Regarding the --verbose option, I failed to RTFM. :-)

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:31 GMT_

---

LOL, I just learned a new acronym ;)

Yes, I understood the code, but was curious if it would not fail if I do the following:
`basisproject --responsefile my.rsp --doc`
where the `--doc` option is defined only through the template configuration file.

Instead of `parser.parse_args`, it should be `parser.parse_known_args` I believe. To be safe, I just left it out for the time being to focus on my actual task. Just wanted to worry about it separately once that is set up.

I am about to enable it with the next commit after testing it as well.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:33 GMT_

---

Good point.

Yes, the response file seemed to be disabled, but there appears to be another issue. No matter how I run it, if there are any parameters I get an error.

example:
○ ~/basis/test/basis-install/bin/basisproject -v
Could not find a BasisProject.cmake file in /Users/ahundt/Source_Code/coretech/basis/test! Are you sure this is a BASIS project?

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 00:57 GMT_

---

it looks like there is a bug when argv == 1, unusual output will be printed.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 01:01 GMT_

---

When you call the `basisproject` tool without any arguments, you get the help. When you call it with `-h` or `--version`, you get the requested information and it quits. Otherwise, if you do not provide a project name, it assumes you want to _upgrade_ a previously created BASIS-project. This means that it performs a 3-way diff as described in the documentation. If any of the template files have been edited, e.g., because of a newer template version with slight changes, these are merged with the project files.

If you want to create a new project into an empty directory, you must specify the `--name` and `--description` options.

See also http://www.rad.upenn.edu/sbia/software/basis/howto/create-and-modify-project.html#modify-an-existing-project and http://www.rad.upenn.edu/sbia/software/basis/howto/create-and-modify-project.html#update-a-project.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 01:05 GMT_

---

Response file feature enabled (with slight modifications) in c3a817f75c0ced996fb1b1d89d768ee3cb0c4165.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 01:12 GMT_

---

Hmm, do your changes mean that --responsefile cannot be used from within a response file?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 01:16 GMT_

---

I actually had not tested this. You would need to use the `@/path/to/file` syntax instead. I would actually prefer to have a different include syntax, but also understand that `--responsefile` is just another option. I will make the needed change.

BTW, where is the origin of the `@/path/to/file` syntax ? Is it common on Windows ?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 01:25 GMT_

---

Should be fixed by 4fa12f5481452f3708933fc150b0576acd4ad1af.

BTW I noticed that the template options such as `--nodoc` or `--nosrc` seem to be ignored at the moment. This might be related to the issue you reported here. I have a look.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 01:37 GMT_

---

Working fine now after commit 45a25692fb6b25f4b2a05bdbf894fb46e757eda5. Thus far only tested by specifying the `data/template-1.0` template explicitly using the `--template` option. This worked fine and the options of the more recent templates were not available as it should be.

Can you share your project template or parts of it via email ? Then I can take a look at your issue if it is not resolved by any of the recent fixes.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 02:07 GMT_

---

I'm not actually sure where it is from, but I think I've seen it before.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 02:22 GMT_

---

I guess here ? http://msdn.microsoft.com/en-us/library/3te4xt0y.aspx

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 02:36 GMT_

---

I bet that's it

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 13, 2013 at 02:41 GMT_

---

Thanks for explaining the upgrade part. That makes a lot more sense. How would you feel if I added your explanation to "Are you sure this is a BASIS project?"?

I think it would be a great debug message if it looked like this:  

```
Could not find a BasisProject.cmake file in [path]! Are you sure this is a BASIS project? 

If you do not provide a project name, it assumes you want to upgrade a previously created BASIS-project. 
This means that it performs a 3-way diff as described in the documentation. If any of the template files 
have been edited, e.g., because of a newer template version with slight changes, these are merged with 
the project files.

If you want to create a new project into an empty directory, you must specify the --name and --description options.

See also http://www.rad.upenn.edu/sbia/software/basis/howto/create-and-modify-project.html#modify-an-existing-project and http://www.rad.upenn.edu/sbia/software/basis/howto/create-and-modify-project.html#update-a-project.
```

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 02:46 GMT_

---

Sounds good. But I would leave out the URLs as they may change. See user
manual should be sufficient next to your explanations.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 18, 2013 at 23:20 GMT_

---

added in changeset: d402cec   merged in: f3d4cfe 
I've tested that this is working.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Wednesday Nov 27, 2013 at 02:50 GMT_

---

Documentation of this issue is covered by #231 and #248, marking completed.
